### PR TITLE
Detect version

### PR DESCRIPTION
**Pull request by hoangdat**

### CODE REVIEW

**Comment by hoangdat** [comment]
login with v4:

**Comment by hoangdat** [comment]
login with v5:


https://user-images.githubusercontent.com/6462404/150092809-e847d12f-dd81-4d21-a82c-db7d923e5299.mov

**Comment by hoangdat** [comment]
download: 


https://user-images.githubusercontent.com/6462404/150093196-bb077e87-7096-4ced-a4f1-b3cc958ee4c8.mov

**Comment by hoangdat** [comment]
upload:


https://user-images.githubusercontent.com/6462404/150093524-898ccac8-9a8d-48c4-b325-f1aae72bf88e.mov

**Comment by hoangdat** [comment]
oidc:


https://user-images.githubusercontent.com/6462404/150093707-3f0100ed-65fc-43d5-a6e2-f16bd048eff6.mov

File: data/lib/src/datasource_impl/authentication_oidc_datasource_impl.dart [hunk 1]
```diff
@@ -94,18 +95,22 @@ class AuthenticationOIDCDataSourceImpl implements AuthenticationOIDCDataSource {
...
+          apiVersion,
           tokenOIDC.token,
           PermanentTokenBodyRequest('Token-${deviceManager.getPlatformString()}-$deviceName'),
           otpCode: otpCode);
       return permanentToken.toToken();
     }).catchError((error) {
       remoteExceptionThrower.throwRemoteException(error, handler: (DioError error) {
-        if (error.response?.statusCode == 401) {
+        if (error.response?.statusCode == 404) {
+          developer.log('createPermanentTokenWithOIDC(): $error', name: 'AuthenticationOIDCDataSourceImpl');
```
Comment (dab246) [inline data/lib/src/datasource_impl/authentication_oidc_datasource_impl.dart:111]: remove it

File: domain/lib/src/usecases/authentication/create_permanent_token_oidc_interactor.dart [hunk 2]
```diff
@@ -29,25 +29,50 @@
...
-    try {
-      final token = await authenticationOIDCRepository.createPermanentTokenWithOIDC(baseUrl, tokenOIDC, otpCode: otpCode);
-      await tokenRepository.persistToken(token);
-      await credentialRepository.saveBaseUrl(baseUrl);
-      return Right(AuthenticationViewState(token));
-    } catch (e) {
-      return Left(AuthenticationFailure(e));
+    for (var i = 0; i <= APIVersionSupported.values.length; i++) {
+      try {
+        developer.log('execute(): $baseUrl in ${APIVersionSupported.values[i]}', name: 'CreatePermanentTokenOIDCInteractor');
```
Comment (dab246) [inline domain/lib/src/usecases/authentication/create_permanent_token_oidc_interactor.dart:54]: remove it

File: domain/lib/src/usecases/authentication/create_permanent_token_oidc_interactor.dart [hunk 3]
```diff
@@ -29,25 +29,50 @@
...
+        final token = await authenticationOIDCRepository.createPermanentTokenWithOIDC(
+            baseUrl,
+            APIVersionSupported.values[i],
+            tokenOIDC,
+            otpCode: otpCode);
+        await tokenRepository.persistToken(token);
+        await credentialRepository.saveBaseUrl(baseUrl);
+        developer.log('execute(): detected version ${APIVersionSupported.values[i]}', name: 'CreatePermanentTokenOIDCInteractor');
```
Comment (dab246) [inline domain/lib/src/usecases/authentication/create_permanent_token_oidc_interactor.dart:62]: remove it

**Review by dab246 — COMMENTED** [review]
- Remove all debug logger
- I have a question. Is it possible to have more than 1 version supported at the same time?. Here you use a loop to get the supported version. The current version list is
````
enum APIVersionSupported {
   v5, v4, v2
}
````
If v5 and v4 are both supported. By default, the first value of the v5 loop will return 200, so the user will use v5. How can I use v4 when v5 already returns 200.

**Review by dab246 — APPROVED** [review]

**Comment by hoangdat** [comment]
> * I have a question. Is it possible to have more than 1 version supported at the same time?. Here you use a loop to get the supported version. The current version list is
> 
> ```
> enum APIVersionSupported {
>    v5, v4, v2
> }
> ```
> 
> If v5 and v4 are both supported. By default, the first value of the v5 loop will return 200, so the user will use v5. How can I use v4 when v5 already returns 200.

yes, we prefer to use newer version in back-end. Fx: in Back-end `v5`, we should use API `v5`.

**Pull request merged**